### PR TITLE
Refactor request context

### PR DESCRIPTION
**Pull request by dr3**
**Overall change:** Refactor request context to be self contained

**Code changes:**

- Moves logic from page handler to request context
- Updates usages in tests and stories around simorgh

Zero changes expected. Likely regression surrounds things like the amp/canonical logic, ATI firing 

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [ ] Test engineer approval

### CODE REVIEW

File: src/app/Layouts/__snapshots__/defaultPageWrapper.test.jsx.snap [hunk 1]
```diff
@@ -7,14 +7,11 @@ exports[`defaultPageWrapper should render with children 1`] = `
   >
     <GlobalStyle />
     <RequestContextProvider
-      env="live"
+      bbcOrigin="https://www.bbc.com"
```
Comment (ghost) [inline src/app/Layouts/__snapshots__/defaultPageWrapper.test.jsx.snap]: Why the change to `bbcOrigin`?

Comment (dr3) [reply src/app/Layouts/__snapshots__/defaultPageWrapper.test.jsx.snap]: Its not changing it, but rather passing it to `RequestContextProvider` to determine the `origin` prop that was previously passed in

File: src/app/contexts/RequestContext/getEnv/index.test.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import getEnv from './getEnv';
+import getEnv from './index';
```
Comment (sareh) [inline src/app/contexts/RequestContext/getEnv/index.test.js]:
```suggestion
import getEnv from '.';
```

File: src/app/contexts/RequestContext/getStatsDestination/index.test.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import getStatsDestination from './getStatsDestination';
+import getStatsDestination from './index';
```
Comment (sareh) [inline src/app/contexts/RequestContext/getStatsDestination/index.test.js]:
```suggestion
import getStatsDestination from '.';
```

File: src/app/contexts/RequestContext/getStatsPageIdentifier/index.test.js [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import getStatsPageIdentifier from './getStatsPageIdentifier';
+import getStatsPageIdentifier from './index';
```
Comment (sareh) [inline src/app/contexts/RequestContext/getStatsPageIdentifier/index.test.js]:
```suggestion
import getStatsPageIdentifier from '.';
```

File: src/app/contexts/RequestContext/index.test.jsx [hunk 5]
```diff
@@ -1,102 +1,110 @@
...
-    isUK: false,
-    origin: 'https://www.bbc.com',
+jest.mock('./getStatsDestination');
+jest.mock('./getStatsPageIdentifier');
+jest.mock('./getOriginContext');
+jest.mock('./getEnv');
+
+getStatsDestination.default.mockReturnValue('getStatsDestination');
+getStatsPageIdentifier.default.mockReturnValue('getStatsPageIdentifier');
+getOriginContext.default.mockReturnValue({ isUK: 'isUK', origin: 'origin' });
```
Comment (sareh) [inline src/app/contexts/RequestContext/index.test.jsx]: Ensuring that we're returning the same types as the RequestContext expects:

```suggestion
getOriginContext.default.mockReturnValue({ isUK: true, origin: 'origin' });
```

File: src/app/contexts/RequestContext/index.test.jsx [hunk 6]
```diff
@@ -1,102 +1,110 @@
...
+  id: 'id',
+  isAmp: true,
+  pageType: 'frontpage',
+  service: 'service',
+};
+
+const expectedOutput = {
+  env: 'getEnv',
+  id: input.id,
+  isUK: 'isUK',
```
Comment (sareh) [inline src/app/contexts/RequestContext/index.test.jsx]: Should be same type as that expected by the Context
```suggestion
  isUK: true
```

File: src/app/contexts/RequestContext/index.test.jsx [hunk 7]
```diff
@@ -1,102 +1,110 @@
...
+  it('should return expected values', () => {
+    render(
+      <RequestContextProvider {...input}>
+        <Component />
+      </RequestContextProvider>,
+    );
+
+    expect(getStatsDestination.default).toHaveBeenCalledWith({
+      env: 'getEnv',
+      isUK: 'isUK',
```
Comment (sareh) [inline src/app/contexts/RequestContext/index.test.jsx]:
```suggestion
      isUK: true,
```

File: src/app/containers/ATIAnalytics/ArticleAtiParams.test.jsx [hunk 8]
```diff
@@ -39,12 +39,11 @@ describe('ArticleAtiParams', () => {
     atiAnalyticsAppName: 'news',
   };
   const requestContextStub = {
-    isUK: true,
-    origin: 'https://www.bbc.co.uk',
+    id: 'c0000000000o',
```
Comment (sareh) [inline src/app/containers/ATIAnalytics/ArticleAtiParams.test.jsx]: The order of the props change depending on the file - could you please ensure they're alphabetically sorted? 
I will separately raise an issue to add the lint rule https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-sort-props.md to our config so that it is mentioned prior to committing in future.

**Review by sareh — CHANGES_REQUESTED** [review]
Looks good so far. I've added comments regarding tests and ordering of props.

File: src/app/contexts/RequestContext/index.test.jsx (hunk 5 above)
Comment (dr3) [reply src/app/contexts/RequestContext/index.test.jsx]: This is simply to ensure the data is passed as expected, all values are their name

Comment (dr3) [reply src/app/contexts/RequestContext/index.test.jsx]: Done :)

File: src/app/contexts/RequestContext/index.test.jsx (hunk 6 above)
Comment (dr3) [reply src/app/contexts/RequestContext/index.test.jsx]: Done :)

File: src/app/containers/ATIAnalytics/ArticleAtiParams.test.jsx (hunk 8 above)
Comment (dr3) [reply src/app/containers/ATIAnalytics/ArticleAtiParams.test.jsx]: Done :)

File: src/app/contexts/RequestContext/index.test.jsx (hunk 7 above)
Comment (dr3) [reply src/app/contexts/RequestContext/index.test.jsx]: Done :)

File: src/app/Layouts/defaultPageWrapper.jsx [hunk 9]
```diff
@@ -7,49 +7,29 @@ import ServiceWorkerContainer from '../containers/ServiceWorker';
...
-          <HeaderContainer />
-          {children}
-          <FooterContainer />
-        </RequestContextProvider>
-      </ServiceContextProvider>
-    </Fragment>
-  );
-};
+const PageWrapper = ({ bbcOrigin, children, id, service, isAmp, pageType }) => (
+  <Fragment>
```
Comment (ghost) [inline src/app/Layouts/defaultPageWrapper.jsx]: I don't think its necessary to wrap this in a fragment.

File: src/app/containers/ArticleMain/index.test.jsx [hunk 10]
```diff
@@ -11,13 +11,11 @@ articleDataNewsNoHeadline.content.model.blocks.shift();
...
+    bbcOrigin="https://www.test.bbc.co.uk"
     id="c0000000000o"
-    isUK
-    origin="https://www.test.bbc.co.uk"
+    isAmp={false}
     pageType="article"
-    platform="canonical"
-    statsDestination="NEWS_PS_TEST"
-    statsPageIdentifier="news.articles.c0000000000o.page"
+    service="news"
```
Comment (ghost) [inline src/app/containers/ArticleMain/index.test.jsx]: Get `service` from props to fix shallow snapshot on the `persian` test.

File: src/app/containers/ArticleMain/__snapshots__/index.test.jsx.snap [hunk 11]
```diff
@@ -157,14 +154,11 @@ exports[`ArticleMain should render a news article correctly 1`] = `
...
+  bbcOrigin="https://www.test.bbc.co.uk"
   id="c0000000000o"
-  isUK={true}
-  origin="https://www.test.bbc.co.uk"
+  isAmp={false}
   pageType="article"
-  platform="canonical"
-  statsDestination="NEWS_PS_TEST"
-  statsPageIdentifier="news.articles.c0000000000o.page"
+  service="news"
```
Comment (ghost) [inline src/app/containers/ArticleMain/__snapshots__/index.test.jsx.snap]:
```suggestion
  service="persian"
```

File: src/app/containers/FrontPageSection/index.stories.jsx [hunk 12]
```diff
@@ -11,12 +11,10 @@ import { RequestContextProvider } from '../../contexts/RequestContext';
...
-      platform={platform}
-      isUK={false}
-      origin={`https://www.bbc.com/${platform}`}
-      pageType="article"
-      statsDestination="WS_NEWS_LANGUAGES_TEST"
-      statsPageIdentifier={`${service}.page`}
+      bbcOrigin="https://www.test.bbc.com"
+      isAmp={platform === 'amp'}
+      pageType="frontpage"
+      service="news"
```
Comment (ghost) [inline src/app/containers/FrontPageSection/index.stories.jsx]: Get `service` from param.

File: src/app/containers/StoryPromo/index.test.jsx [hunk 13]
```diff
@@ -112,12 +112,11 @@ const fixtures = {
...
-      platform={platform || 'canonical'}
-      isUK
-      origin="https://www.bbc.co.uk"
+      bbcOrigin="https://www.test.bbc.co.uk"
       id="c0000000000o"
-      statsDestination="NEWS_PS_TEST"
-      statsPageIdentifier="news.articles.c0000000000o"
+      isAmp={platform === 'amp'}
+      pageType="article"
+      service="news"
```
Comment (ghost) [inline src/app/containers/StoryPromo/index.test.jsx]: Get `service` from `props`

File: src/app/contexts/RequestContext/index.jsx [hunk 14]
```diff
@@ -31,18 +46,15 @@ export const RequestContextProvider = ({
 };
 
 RequestContextProvider.propTypes = {
+  bbcOrigin: string,
   children: node.isRequired,
-  env: string,
   id: string,
+  isAmp: bool.isRequired,
```
Comment (ghost) [inline src/app/contexts/RequestContext/index.jsx:52]: Just thinking, it would be nice if `isAmp` was optional and had a default of `false`.
```jsx
// we would have
<RequestContextProvider />
// instead of
<RequestContextProvier isAmp={false} />
```
If `isAmp` was made required so that we can be explicit about it while using `RequestContextProvider` you can safely ignore my comment.

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/app/containers/ArticleMain/index.test.jsx (hunk 10 above)
Comment (dr3) [reply src/app/containers/ArticleMain/index.test.jsx]: Done :)

File: src/app/containers/ArticleMain/__snapshots__/index.test.jsx.snap (hunk 11 above)
Comment (dr3) [reply src/app/containers/ArticleMain/__snapshots__/index.test.jsx.snap]: Done :)

File: src/app/Layouts/defaultPageWrapper.jsx (hunk 9 above)
Comment (dr3) [reply src/app/Layouts/defaultPageWrapper.jsx]: Done :)

File: src/app/containers/FrontPageSection/index.stories.jsx (hunk 12 above)
Comment (dr3) [reply src/app/containers/FrontPageSection/index.stories.jsx]: Done :)

File: src/app/containers/StoryPromo/index.test.jsx (hunk 13 above)
Comment (dr3) [reply src/app/containers/StoryPromo/index.test.jsx]: Done :)

File: src/app/contexts/RequestContext/index.jsx (hunk 14 above)
Comment (dr3) [reply src/app/contexts/RequestContext/index.jsx:52]: Yeah isAmp is passed from high up in the app, so we should keep it required ideally, even though it does make it a bit more annoying in tests

**Review by pjlee11 — APPROVED** [review]
Looks like it makes sense

**Review by ghost — APPROVED** [review]

**Review by chris-hinds — APPROVED** [review]

**Comment by chris-hinds** [comment]
Approved, but it does have some conflicts now

**Comment by j-pendlebury** [comment]
More conflicts 👎

**Comment by j-pendlebury** [comment]
@sareh are you happy with the changes?

**Review by sareh — APPROVED** [review]
👍 Looks great! ✨

**Comment by david-boydell** [comment]
Amp/Canonical & ATI are working as expected. Can test is UK/is Not UK once this goes up to Test/Stage. This is ready for merge.
